### PR TITLE
Changeable Max Stack Size for Items and Inventory

### PR DESCRIPTION
**Pull request by NotSoDelayed**
### Description
This PR adds support of changeable max stack size of inventories and the recently introduced `max stack size` ItemMeta in Minecraft 1.20.5.

Javadocs:
- (ItemMeta) https://hub.spigotmc.org/javadocs/spigot/org/bukkit/inventory/meta/ItemMeta.html#setMaxStackSize(java.lang.Integer)
- (Inventory) https://hub.spigotmc.org/javadocs/spigot/org/bukkit/inventory/Inventory.html#setMaxStackSize(int)

#### Note
Using GitHub actions to test the attached test scripts due to Skript testing environment abruptly failing. My sincerely apologies for any bulk commits from fixing tests.

---
**Target Minecraft Versions:** Minecraft 1.20.5 (changeable itemstack max stack size)
**Requirements:** Spigot 1.20.5 (changeable max stack size ItemMeta)
**Related Issues:** #5020 (inventory)

### CODE REVIEW

**Comment by cheeezburga** [comment]
Should this PR also include changes to the `is stackable` condition to properly reflect this new stuff?

**Comment by NotSoDelayed** [comment]
> Should this PR also include changes to the `is stackable` condition to properly reflect this new stuff?

The condition checks if the max stack size is more than 1, which to me it looks like it is fine as is, unless I’m missing something.
image

**Comment by cheeezburga** [comment]
> > Should this PR also include changes to the `is stackable` condition to properly reflect this new stuff?
> 
> The condition checks if the max stack size is more than 1, which to me it looks like it is fine as is, unless I’m missing something. image

Ye I have no idea what I was thinking when I looked at the condition earlier, looks so fine. Probably a result of no sleep tbh, my bad.

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 1]
```diff
@@ -18,35 +18,131 @@
...
+	"Since MC 1.20.5 onwards, the maximum stack size of items can be changed (any integer from 1 to 99), and the said item can be stacked up to the set value, up to the maximum stack size of an inventory."
+})
+@Examples({
+	"send \"You can only pick up %max stack size of player's tool% of %type of (player's tool)%\" to player",
+	"",
+	"set the maximum stack size of inventory of all players to 16",
+	"",
+	"add 8 to the maximum stack size of player's tool",
+	"",
+	"reset the maximum stack size of {_gui}"
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:50]:
```suggestion
	"send \"You can only pick up %max stack size of player's tool% of %type of (player's tool)%\" to player",
	"set the maximum stack size of inventory of all players to 16",
	"add 8 to the maximum stack size of player's tool",
	"reset the maximum stack size of {_gui}"
```

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 2]
```diff
@@ -18,35 +18,131 @@
...
+	@Nullable
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		switch (mode) {
+            case ADD:
+			case DELETE:
+            case REMOVE:
+			case RESET:
+			case SET:
+				if (!CHANGEABLE_ITEM_STACK_SIZE && ItemType.class.isAssignableFrom(getExpr().getReturnType())) {
+					Skript.error("Changers for maximum stack size of items requires Minecraft 1.20.5 or newer");
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:85]:
```suggestion
					Skript.error("Changing the maximum stack size of items requires Minecraft 1.20.5 or newer!");
```

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 3]
```diff
@@ -18,35 +18,131 @@
...
+                switch (mode) {
+                    case ADD:
+						size += change;
+                        break;
+                    case SET:
+						size = change;
+						break;
+                    case REMOVE:
+						size -= change;
+                        break;
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:114]: Reset and Delete are not handled.

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 4]
```diff
@@ -18,35 +18,131 @@
...
 	}
-	
-	@SuppressWarnings("null")
+
+	@Override
+	@Nullable
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		switch (mode) {
+            case ADD:
+			case DELETE:
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:80]: Delete is not an appropriate changer here.

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java (hunk 3 above)
Comment (NotSoDelayed) [reply src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:114]: Here where it handles:
https://github.com/SkriptLang/Skript/blob/9e5833800a5d37ff70d7b2413de0bb40a32caaa6/src/main/java/ch/njol/skript/expressions/ExprMaxStack.java#L98-L99

Then here:
https://github.com/SkriptLang/Skript/blob/9e5833800a5d37ff70d7b2413de0bb40a32caaa6/src/main/java/ch/njol/skript/expressions/ExprMaxStack.java#L118

Comment (sovdeeth) [reply src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:114]: Ah I see, I missed the ternary

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 5]
```diff
@@ -18,35 +18,126 @@
...
+				return CollectionUtils.array(Integer.class);
+			default:
+				return null;
+        }
+	}
+
 	@Override
-	public Long convert(final ItemType i) {
-		return (long) i.getRandom().getMaxStackSize();
+	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:91]:
```suggestion
	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
```

File: src/test/skript/tests/syntaxes/expressions/ExprMaxStack.sk [hunk 6]
```diff
@@ -2,3 +2,49 @@ test "max stack":
 	assert max stack size of diamond sword is 1 with "diamond sword max stack size failed"
 	assert max stack size of bucket is 16 with "bucket max stack size failed"
 	assert max stack size of dirt is 64 with "dirt max stack size failed"
+
+test "max stack override - itemtype" when running minecraft "1.20.5":
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/expressions/ExprMaxStack.sk:6]: I'd also like to see some tests of changing multiple items or category item types (all logs) at a time, with varying default sizes. Plus edge cases tests of null values on both sides.

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 7]
```diff
@@ -18,35 +18,126 @@
...
+		register(ExprMaxStack.class, Integer.class, "max[imum] stack[[ ]size]", "itemtypes/inventories");
+	}
+
+	private static final boolean CHANGEABLE_ITEM_STACK_SIZE = Skript.methodExists(ItemMeta.class, "setMaxStackSize", Integer.class);
+
+	@Override
+	@Nullable
+	public Integer convert(Object source) {
+		if (source instanceof ItemType) {
+			return ((ItemType) source).getRandom().getMaxStackSize();
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:63]: It is possible that getRandom returns null for Materials that cannot be represented by an ItemStack. So, if getRandom returns null, you should instead use `itemType.getMaterial().getMaxStackSize();` (defer to the material's value)

**Review by sovdeeth — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 8]
```diff
@@ -1,57 +1,131 @@
...
+		register(ExprMaxStack.class, Integer.class, "max[imum] stack[[ ]size]", "itemtypes/inventories");
 	}
-	
-	@SuppressWarnings("null")
+
+	private static final boolean CHANGEABLE_ITEM_STACK_SIZE = Skript.methodExists(ItemMeta.class, "setMaxStackSize", Integer.class);
+
+	@Override
+	@Nullable
+	public Integer convert(Object source) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:46]: Can use pattern matching for the instanceof checks now since we're on Java 17

**Review by APickledWalrus — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 9]
```diff
@@ -1,57 +1,131 @@
...
+		if (source instanceof ItemType) {
+			Object itemType = ((ItemType) source).getRandomStackOrMaterial();
+			if (itemType instanceof ItemStack)
+				return ((ItemStack) itemType).getMaxStackSize();
+			return ((Material) itemType).getMaxStackSize();
+		} else if (source instanceof Inventory) {
+			return (((Inventory) source).getMaxStackSize());
```
Comment (Moderocky) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:53]:
```suggestion
		if (source instanceof ItemType type) {
			Object itemType = type.getRandomStackOrMaterial();
			if (itemType instanceof ItemStack stack)
				return stack.getMaxStackSize();
			return ((Material) itemType).getMaxStackSize();
		} else if (source instanceof Inventory inventory) {
			return inventory.getMaxStackSize();
```

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 10]
```diff
@@ -1,57 +1,131 @@
...
+
+	@Override
+	@Nullable
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		switch (mode) {
+            case ADD:
+            case REMOVE:
+			case RESET:
+			case SET:
+				if (!CHANGEABLE_ITEM_STACK_SIZE && ItemType.class.isAssignableFrom(getExpr().getReturnType())) {
```
Comment (Moderocky) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:68]:
```suggestion
            case ADD, REMOVE, RESET, SET:
				if (!CHANGEABLE_ITEM_STACK_SIZE && ItemType.class.isAssignableFrom(getExpr().getReturnType())) {
```

**Review by Moderocky — CHANGES_REQUESTED** [review]
We can just tweak these for Java 17 :)

**Comment by NotSoDelayed** [comment]
Java 17 ready!

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 11]
```diff
@@ -1,62 +1,126 @@
...
+	private static final boolean CHANGEABLE_ITEM_STACK_SIZE = Skript.methodExists(ItemMeta.class, "setMaxStackSize", Integer.class);
+
 	@Override
-	public Long convert(ItemType itemType) {
-		Object random = itemType.getRandomStackOrMaterial();
-		if (random instanceof Material)
-			return (long) ((Material) random).getMaxStackSize();
-		return (long) ((ItemStack) random).getMaxStackSize();
+	@Nullable
+	public Integer convert(Object from) {
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:46]:
```suggestion
	public @Nullable Integer convert(Object from) {
```

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 12]
```diff
@@ -1,62 +1,126 @@
...
+            default -> null;
+        };
+	}
+
+	@Override
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
+		for (Object source : getExpr().getArray(event)) {
+			Integer change = null;
+			if (mode != ChangeMode.RESET)
+				change = (int) delta[0];
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:74]: You can move this outside the for loop so it only has to be got once

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 13]
```diff
@@ -1,62 +1,126 @@
...
+                    case ADD:
+						size += change;
+                        break;
+                    case SET:
+						size = change;
+						break;
+                    case REMOVE:
+						size -= change;
+                        break;
+                }
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:89]: This can be an enhanced switch, like the one below with the `->` stuff

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 14]
```diff
@@ -1,62 +1,126 @@
...
- * @author joeuguce99
- */
 @Name("Maximum Stack Size")
-@Description("The maximum stack size of the specified material, e.g. 64 for torches, 16 for buckets, and 1 for swords.")
-@Examples("send \"You can only pick up %max stack size of player's tool% of %type of (player's tool)%\" to player")
-@Since("2.1")
-public class ExprMaxStack extends SimplePropertyExpression<ItemType, Long> {
+@Description({
+	"The maximum stack size of an item (e.g. 64 for torches, 16 for buckets, and 1 for swords), or of an inventory to have items to stack up to.",
+	"Since MC 1.20.5 onwards, the maximum stack size of items can be changed (any integer from 1 to 99), and the said item can be stacked up to the set value, up to the maximum stack size of an inventory."
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:26]:
```suggestion
	"The maximum stack size of an item (e.g. 64 for torches, 16 for buckets, 1 for swords, etc.) or inventory.",
	"In 1.20.5+, the maximum stack size of items can be changed to any integer from 1 to 99, and stacked up to the maximum stack size of the inventory they're in."
```

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 15]
```diff
@@ -1,62 +1,126 @@
...
+})
+@Examples({
+	"send \"You can only pick up %max stack size of player's tool% of %type of (player's tool)%\" to player",
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:29]: This seems like a weird example to use. They can still pick up more than the max stack size of the item right?

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 16]
```diff
@@ -1,62 +1,126 @@
...
+	"set the maximum stack size of inventory of all players to 16",
+	"add 8 to the maximum stack size of player's tool",
+	"reset the maximum stack size of {_gui}"
+})
+@Since("2.1, INSERT VERSION (changeable, inventory support)")
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:34]:
```suggestion
@Since("2.1, INSERT VERSION (changeable, inventories)")
```

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 17]
```diff
@@ -1,62 +1,126 @@
...
+@RequiredPlugins("Spigot 1.20.5+ (changeable item max stack size)")
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:35]:
```suggestion
@RequiredPlugins("Spigot 1.20.5+ (changeable)")
```

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 18]
```diff
@@ -1,62 +1,126 @@
...
+		if (from instanceof Inventory inventory)
+			return inventory.getMaxStackSize();
+		return null;
 	}
 
 	@Override
-	public Class<? extends Long> getReturnType() {
-		return Long.class;
+	@Nullable
+	public Class<?>[] acceptChange(ChangeMode mode) {
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:56]:
```suggestion
	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
```

**Review by cheeezburga — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java (hunk 13 above)
Comment (NotSoDelayed) [reply src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:89]: Oops how did I missed out this old switch…

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java (hunk 15 above)
Comment (NotSoDelayed) [reply src/main/java/ch/njol/skript/expressions/ExprMaxStack.java]: Good catch. Any more than that will be carried over to the next slot.

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 19]
```diff
@@ -1,62 +1,118 @@
...
+			return getMaxStackSize(itemType);
+		if (from instanceof Inventory inventory)
+			return inventory.getMaxStackSize();
+		return null;
 	}
 
 	@Override
-	public Class<? extends Long> getReturnType() {
-		return Long.class;
+	public @Nullable Class<?>[] acceptChange(ChangeMode mode) {
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:54]:
```suggestion
	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
```

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 20]
```diff
@@ -1,62 +1,118 @@
...
+            }
+            default -> null;
+        };
+	}
+
+	@Override
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
+		Integer change = null;
+		if (mode != ChangeMode.RESET)
+			change = (int) delta[0];
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:71]: This part is kinda weird to me. I usually just do this instead, which would mean some of the following logic would have to change too (mainly just where you check if `change != null` I think)
```suggestion
		int change = delta == null ? 0 : ((Number) delta[0]).intValue();
```

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 21]
```diff
@@ -1,62 +1,118 @@
...
 	protected String getPropertyName() {
 		return "maximum stack size";
 	}
 
+	private static int getMaxStackSize(ItemType itemType) {
+		Object item = itemType.getRandomStackOrMaterial();
+		if (item instanceof ItemStack stack)
+			return stack.getMaxStackSize();
+		if (item instanceof Material material)
+			return material.getMaxStackSize();
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:114]:
```suggestion
		if (item instanceof ItemStack stack) {
			return stack.getMaxStackSize();
		} else if (item instanceof Material material) {
			return material.getMaxStackSize();
		}
```

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 22]
```diff
@@ -1,62 +1,118 @@
...
-	public Long convert(ItemType itemType) {
-		Object random = itemType.getRandomStackOrMaterial();
-		if (random instanceof Material)
-			return (long) ((Material) random).getMaxStackSize();
-		return (long) ((ItemStack) random).getMaxStackSize();
+	public @Nullable Integer convert(Object from) {
+		if (from instanceof ItemType itemType)
+			return getMaxStackSize(itemType);
+		if (from instanceof Inventory inventory)
+			return inventory.getMaxStackSize();
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:49]:
```suggestion
		if (from instanceof ItemType itemType) {
			return getMaxStackSize(itemType);
		} else if (from instanceof Inventory inventory) {
			return inventory.getMaxStackSize();
		}
```

**Review by cheeezburga — CHANGES_REQUESTED** [review]
Just some small things

**Review by cheeezburga — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java [hunk 23]
```diff
@@ -1,62 +1,116 @@
...
+	"In 1.20.5+, the maximum stack size of items can be changed to any integer from 1 to 99, and stacked up to the maximum stack size of the inventory they're in."
+})
+@Examples({
+	"send \"You can hold %max stack size of player's tool% of %type of player's tool% in a slot.\" to player",
+	"set the maximum stack size of inventory of all players to 16",
+	"add 8 to the maximum stack size of player's tool",
+	"reset the maximum stack size of {_gui}"
+})
+@Since("2.1, INSERT VERSION (changeable, inventories)")
+@RequiredPlugins("Spigot 1.20.5+ (changeable)")
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:35]:
```suggestion
@RequiredPlugins("Minecraft 1.20.5+ (changeable)")
```

**Review by Efnilite — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprMaxStack.java (hunk 23 above)
Comment (NotSoDelayed) [reply src/main/java/ch/njol/skript/expressions/ExprMaxStack.java:35]: Both term `Minecraft` and `Spigot` are acceptable based on past debates between members — unless a member objects, I’m marking this as resolved
